base/abstractarray.jl

```diff
@@ -86,8 +128,48 @@ linearindexing(A::AbstractArray, B::AbstractArray...) = linearindexing(linearind
 linearindexing(::LinearFast, ::LinearFast) = LinearFast()
 linearindexing(::LinearIndexing, ::LinearIndexing) = LinearSlow()
 
+abstract IndicesBehavior
+immutable IndicesStartAt1  <: IndicesBehavior end   # indices 1:size(A,d)
+immutable IndicesUnitRange <: IndicesBehavior end   # arb UnitRange indices
+immutable IndicesList      <: IndicesBehavior end   # indices like (:cat, :dog, :mouse)
+
+indicesbehavior(A::AbstractArray) = indicesbehavior(typeof(A))
+indicesbehavior{T<:AbstractArray}(::Type{T}) = IndicesStartAt1()
+indicesbehavior(::Number) = IndicesStartAt1()
+
+abstract IndicesPerformance
+immutable IndicesFast1D <: IndicesPerformance end  # indices(A, d) is fast
+immutable IndicesSlow1D <: IndicesPerformance end  # indices(A) is better than indices(A,d)
+
+indicesperformance(A::AbstractArray) = indicesperformance(typeof(A))
+indicesperformance{T<:AbstractArray}(::Type{T}) = IndicesFast1D()
+
+"""
+    shape(A)
+
+Returns a tuple specifying the "shape" of array `A`. For arrays with
+conventional indexing (indices start at 1), this is equivalent to
+`size(A)`; otherwise it is equivalent to `incides(A)`.
+"""
```

Comment:
@JeffBezanson, I see two problems with that plan, both related to performance:

> if you want to be more generic you have to use `length(indices(A,n))`

While working on this PR, one of the unexpected gotchas I encountered was that we [use checked arithmetic](https://github.com/JuliaLang/julia/blob/ab0c98a750c8f09bc502e6b81455603dafb7b80a/base/range.jl#L353-L354) to compute `length` for `UnitRange{Int}`, and in some very performance-sensitive operations (e.g., `sub2ind`) that turns out to be devastating. I introduced the non-exported [`unsafe_length`](https://github.com/JuliaLang/julia/blob/master/base/abstractarray.jl#L1289) to circumvent that. (If you look at the code near that link, you'll notice that `sub2ind` still has two separate implementations for conventional and unconventional arrays: performing two extra arithmetic operations per dimension introduces a modest but easily-measurable performance hit. This kind of thing is why nanosoldier and I became such good friends recently.)

As a consequence (plus the fact that it's more verbose), I'd argue that `length(indices(A,n))` is probably not the syntax we want to encourage folks towards for the long-haul. 

> We might need a release where bounds checks for non-1:n-indexed arrays are never removed.

I *really* like that idea. The one thing that scares me is sociological: given the gains achievable particularly from `@simd` (which works only when bounds-checks have been removed), this becomes a disincentive to use such arrays. That might drag out this transition longer (fewer algorithms will be tested against unconventional arrays). In an ideal world, all packages will be updated over the course of a single release cycle, although I suspect that's unlikely even in the best of worlds.

@eschnett,

> You could define `size` only for 1-based (old-style) arrays, and introduce have two new functions `indices` (as you defined it) and `shape` (as in numpy). Later, `size` can be deprecated.

I think that's a fine plan if we're happy with having `shape` replace `size` everywhere: if we go that route, I'd propose we deprecate `size` for julia-0.5 (why wait?). But, I think `size` is a little more explicit about what this operation means than `shape` is; by that measure migrating to `shape` would be a step backwards.

How about this as a plan: `size(A,n)` continues to work fine for 1-based arrays, and throws an error otherwise. We introduce a `@this_function_certified_as_safe_for_unconventional_arrays` macro that rewrites `size` as `Base.size_ok` (which is what people implement for non-1:n arrays). Obviously this plan requires cooperation from the folks who write offset-array packages, but I suspect such folks would be more likely than most to pay attention to such details (and indeed will appreciate our concern for their welfare :smile:), so I doubt that will be a concern in practice. At some point in the future we go back to using `size` for this operation and remove the need for the macro.

I'll point out that [very similar considerations](https://github.com/JuliaLang/julialang.github.com/blob/52b20d2f1a2c39d0ea8c4abbc05dba2bd4b454e5/blog/_posts/2016-06-12-offset-arrays.md#linear-indexing-linearindices) apply to `length`; I introduced `linearindices` as a safe alternative. In this case I don't see a great alternative to introducing a new export, because there needs to be a difference in behavior for 1-dimensional arrays vs all others. But we might also want to make `length` throw an error for non-1:n arrays, just to be safe (and of course play the same macro trick).

I'll await feedback on this new proposal. Also want to say how much I appreciate this discussion. We may have figured out a way to make code vastly safer while also eliminating two exports (`shape` and `allocate_for`); I'd call that unqualified success!